src/Flysystem/FlysystemAssetStore.php

```diff
@@ -685,7 +691,7 @@ protected function truncateDirectory($dirname, Filesystem $filesystem)
         if ($dirname
             && ltrim($dirname ?? '', '.')
             && !$this->config()->get('keep_empty_dirs')
-            && !$filesystem->listContents($dirname)->toArray()
+            && $filesystem->isEmpty($dirname)
```

Comment:
This single change is where most of the performance gains will be.

With a setup of ~4000 files (including variants, so pretty conservative as these things go when you consider just uploading a file creates 3 variants for a total of 4 files per upload), calling this method 10 times, the old implementation took ~100ms while the new implementation takes ~2ms.

That ~100ms scaled with the number of files you had, while the ~2ms is constant regardless of how many files there are.

Reply:
Note we don't have to do a type check here because the `Filesystem` type in the method signature is explicitly `SilverStripe\Assets\Flysystem\Filesystem`